cellmlmanip/_singularity_fixes.py

```diff
@@ -0,0 +1,252 @@
+                            find_v_up = solveset(u - U_offset, V)
+                            find_v_low = tuple(find_v_low)
+                            find_v_up = tuple(find_v_up)
+                            assert find_v_low and find_v_up and len(find_v_low) == len(find_v_up) == len(sp) == 1, \
+                                'Expecting exactly 1 singularity point '
+                            (Vmin, Vmax, sp) = (find_v_low[0], find_v_up[0], sp[0])
+                    except TypeError:
+                        pass  # Result could be 'ConditionSet' which is not iterable and not Real
+
+                if Vmin is not None:  # check top
+                    for n in num:  # Check arguments in numerator (or denominator)
+                        match = n.match(Z * V - Z * SP_wildcard)  # search for a multiple of V - sp
+                        found_on_top = check_top_match(match, sp)
+                        if not found_on_top:
+                            # search for a exp(multiple of V - sp)
+                            match = n.match(exp_function(Z * V - Z * SP_wildcard))
+                            found_on_top = check_top_match(match, sp)
+                        if found_on_top:  # We've found a match stop looking in the other numerator arguments
+                            break
+                    if found_on_top:  # found singularity, no need to try further
+                        break
+                    else:
+                        (Vmin, Vmax, sp) = None, None, None
+        if Vmin is not None and found_on_top:  # found singularity, no need to try further
+            break
+        else:
+            (Vmin, Vmax, sp) = None, None, None
+
+    # Put dummies back in and return the singularity point and range bundries
+    return (float_dummies(Vmin), float_dummies(Vmax), float_dummies(sp))
+
+
+def _fix_expr_parts(expr, V, U_offset, exp_function):
+    """
+    Removes fixable singularities and replaces them with piecewise expressions.
+
+    :return: either (Vmin, Vmax, singularity point, expression, True)
+                    if we have identified a singularity needs to be made but it can't be done yet.
+                    For eample if 2 singularities with the same singularity point are added up,
+                    a single singularity is constructed instead.
+             or ``(None, None, None, fixed expr, fixed_expr has piecewise)`` otherwise
+
+    see :meth:`remove_fixable_singularities for more details` "
+    """
+
+    if isinstance(expr, Mul):  # 1 * A --> A (remove unneeded 1 *)
+        expr = Mul(*[a for a in expr.args if not str(a) in ('1.0', '1')])
+
+    # Turn Quantity dummies into numbers, to enable analysis
+    subs_dict = {d: d.evalf(FLOAT_PRECISION) for d in expr.atoms(cellmlmanip.Quantity)}
+    check_U_expr = expr.xreplace(subs_dict)
+
+    if not expr.has(exp_function):  # Expressions without exp don't have GHK-like equations
```

Comment:
More efficient to check this first, before removing 1* and dummy conversion?

Reply:
good point! als by the way `.has(...)` is a lot quicker than `.atoms(..)`